### ISSUE TITLE
Segmentation weird straight lines

### ISSUE THREAD

**Issue by rob-safi** [ultralytics/ultralytics#18326]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I recently trained a yolo11m-seg model and training looks normal and in most images it does really well but in some cases I get weird straight lines. I checked the training data pre and post augmentations and didn't notice anything off.
I have tried with and without retina masks but it doesn't change anything.

Any idea how to mitigate this?

Here are two examples:

![gdrive_inspection_data__IMAGE__40390_IMG_6489](https://github.com/user-attachments/assets/148a22df-f7e5-44d8-9d7a-7ad2bbf95eaa)
![loose_materials_2024_11_20_alec_bywaters_pet__IMAGE__IMG_5467](https://github.com/user-attachments/assets/42d7e794-c604-43bf-b82e-f9ca8cd1fd22)

Here is the full args in case you notice something weird:
```
task: segment
mode: train
model: yolo11m-seg.pt
data: .../v1.yaml
epochs: 100
time: null
patience: 50
batch: -1
imgsz: 600
save: true
save_period: -1
cache: false
device: null
workers: 6
project: .../yolo11m-seg
name: train
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: true
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: false
mask_ratio: 1
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: false
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.4
hsv_v: 0.4
degrees: 0.0
translate: 0.05
scale: 0.1
shear: 0.0
perspective: 0.0001
flipud: 0.5
fliplr: 0.5
bgr: 0.0
mosaic: 0.0
mixup: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.4
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml
save_dir: .../train
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rob-safi, thank you for sharing the details of your issue and for your interest in Ultralytics 🚀! This is an automated response to help you get started — an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com) for guidance, where you can find detailed explanations, including the [segmentation task documentation](https://docs.ultralytics.com/tasks/segment/) and training tips for improving results.

If this is a 🐛 Bug Report, please provide a concise [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that clearly demonstrates the issue. This helps us quickly identify the root cause and debug more effectively.

If this is a custom training ❓ Question, please provide additional context, such as:
- Example images from your dataset showing the problem cases (if permissible and not already included)
- Any unique preprocessing steps or augmentations applied
- Training logs or metrics that may provide useful insight
- Confirm whether you’ve reviewed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

## Troubleshooting

To ensure you're working with the latest updates and features, please upgrade your `ultralytics` package to the latest version:

```bash
pip install -U ultralytics
```

As a general checklist:
1. Confirm that your environment meets all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) (e.g., Python>=3.8 and PyTorch>=1.8).
2. If the issue persists, try running your training process in a verified environment such as the following (each includes pre-installed dependencies):
   - **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
   - **Cloud VMs**: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
   - **Docker**: See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community Support

Connect with the Ultralytics community for additional guidance and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussion
- [Discourse](https://community.ultralytics.com) for in-depth technical Q&A
- [Subreddit](https://reddit.com/r/Ultralytics) to share insights and discuss with other users

## Next Steps

Please let us know if any specific part of your configuration is unclear or causing concern. If you've already tried specific remedies (e.g., using retina masks), sharing those details helps us fine-tune our assistance. 🚀

Thank you for helping us improve, and we’ll begin investigating further with the provided details!

**Comment by Y-T-G** [comment]
What lines are you referring to?

**Comment by rob-safi** [comment]
@Y-T-G on the right of the detection there is an artificial corner that gets created in the first image and in the second image the right of the segmentation is a straight line which seems a bit weird considering it cuts through objects
![image](https://github.com/user-attachments/assets/56d44d4c-b9ac-486f-98a4-65452c58d391)
![image](https://github.com/user-attachments/assets/caa7103b-6311-4748-92de-525703972097)

The training data does not having artifacts like this

**Comment by Y-T-G** [comment]
What if you use `retina_masks=True`?

**Comment by sun510001** [comment]
The same issue occurs when using imgsz=1280 during segmentation training. 
This issue is present in the generated val_batch_pred.jpg during training.

**Comment by Y-T-G** [comment]
What version of Ultralytics is this?

**Comment by Y-T-G** [comment]
Does it occur if you use default `mask_ratio` (4)?

**Comment by Y-T-G** [comment]
I think you should provide all the arguments that are custom, instead of the whole thing because it's hard to tell which are custom and which are default.

**Comment by sun510001** [comment]
> What version of Ultralytics is this?

Tried the latest version and v8.2.103, both versions have the issue.
checked the args.yaml. `mask_ratio: 4`
Results are normal when all arguments are custom, and the issue occurs when using imgsz=1280.

**Comment by Y-T-G** [comment]
Can you upload the pickled files (`pred.pkl`) for the images with the issue?

```python
import pickle

results = model.predict(["problematic_image_1.png", "problematic_image_2.png"], embed=[-1])

with open("pred.pkl", "wb") as f:
   pickle.dump(results, f)
```

**Comment by sun510001** [comment]
> Can you upload the pickled files (`pred.pkl`) for the images with the issue?
> 
> ```python
> import pickle
> 
> results = model.predict(["problematic_image_1.png", "problematic_image_2.png"], embed=[-1])
> 
> with open("pred.pkl", "wb") as f:
>    pickle.dump(results, f)
> ```

here
https://github.com/sun510001/road_seg/raw/refs/heads/master/pred.pkl

**Comment by glenn-jocher** [comment]
Thank you for sharing the file. We'll review the data and investigate the issue further. If we identify any anomalies or solutions, we will follow up here.

**Comment by Y-T-G** [comment]
Does it occur if you add this before your code?

```python
import ultralytics.utils.ops as ops
def crop_mask(masks, boxes):
    """
    It takes a mask and a bounding box, and returns a mask that is cropped to the bounding box.

    Args:
        masks (torch.Tensor): [n, h, w] tensor of masks
        boxes (torch.Tensor): [n, 4] tensor of bbox coordinates in relative point form

    Returns:
        (torch.Tensor): The masks are being cropped to the bounding box.
    """
    _, h, w = masks.shape
    x1, y1, x2, y2 = torch.chunk(boxes[:, :, None].int(), 4, 1)  # x1 shape(n,1,1)
    r = torch.arange(w, device=masks.device, dtype=x1.dtype)[None, None, :]  # rows shape(1,1,w)
    c = torch.arange(h, device=masks.device, dtype=x1.dtype)[None, :, None]  # cols shape(1,h,1)

    return masks * ((r >= x1) * (r <= x2) * (c > y1) * (c < y2))
ops.crop_mask = crop_mask
```

**Comment by sun510001** [comment]
> Does it occur if you add this before your code?
> 
> ```python
> import ultralytics.utils.ops as ops
> def crop_mask(masks, boxes):
>     """
>     It takes a mask and a bounding box, and returns a mask that is cropped to the bounding box.
> 
>     Args:
>         masks (torch.Tensor): [n, h, w] tensor of masks
>         boxes (torch.Tensor): [n, 4] tensor of bbox coordinates in relative point form
> 
>     Returns:
>         (torch.Tensor): The masks are being cropped to the bounding box.
>     """
>     _, h, w = masks.shape
>     x1, y1, x2, y2 = torch.chunk(boxes[:, :, None].int(), 4, 1)  # x1 shape(n,1,1)
>     r = torch.arange(w, device=masks.device, dtype=x1.dtype)[None, None, :]  # rows shape(1,1,w)
>     c = torch.arange(h, device=masks.device, dtype=x1.dtype)[None, :, None]  # cols shape(1,h,1)
> 
>     return masks * ((r >= x1) * (r <= x2) * (c > y1) * (c < y2))
> ops.crop_mask = crop_mask
> ```

yes, I replaced the function crop_mask and retrained the model. The issue still persists.
* labels:
![val_batch0_labels](https://github.com/user-attachments/assets/019c4fc0-2ae3-47da-8310-c6975f325a7c)
* pred: 
![val_batch0_pred](https://github.com/user-attachments/assets/a3cd872b-a993-4144-bf5c-eac90946ecae)

**Comment by Y-T-G** [comment]
I don't see the issue in those images.

Are you referring to the masks being outside of the box? Because that's not occurring in those images.

**Comment by Y-T-G** [comment]
> @Y-T-G on the right of the detection there is an artificial corner that gets created in the first image and in the second image the right of the segmentation is a straight line which seems a bit weird considering it cuts through objects
> ![image](https://github.com/user-attachments/assets/56d44d4c-b9ac-486f-98a4-65452c58d391)
> ![image](https://github.com/user-attachments/assets/caa7103b-6311-4748-92de-525703972097)
> 
> The training data does not having artifacts like this
> 

I thought you were referring to how the mask edges here are outside the box.

**Comment by Y-T-G** [comment]
If you are referring to the boxes getting cropped abruptly, then that is related to the receptive field issue because your objects are too large. You should be using a smaller imgsz.

https://github.com/ultralytics/ultralytics/issues/18008#issuecomment-2517931511

**Comment by rob-safi** [comment]
@Y-T-G I reduced the image size to 300 and still getting these artifacts (see top left corner of the image below for an example)
I feel like reducing more will degrade the images too much, do you have any other ideas?
Can yolo not be used in cases where the object can be the size of the image?

![loose_materials_2024_12_05_gary_bywaters_pet__IMAGE__Copy of IMG_0496](https://github.com/user-attachments/assets/a1d4ce03-401d-465c-b4b2-be3a8e91c607)

**Comment by Y-T-G** [comment]
@rob-safi Are you referring to a part of the mask being outside the box?

Or are you referring to there being mask at that area at all? Because the latter is probably due to the model not performing well. What's the mAP?

**Comment by rob-safi** [comment]
Hi, No I'm referring to the fact that the detection includes part of the white bag (upper and lower left corners of the detection). These are False Positives but the real issue is that the bag is only partially detected and I find it very suspicious that the detection ends in sharp corners when there is nothing to suggest objects should stop here.

mAP50(M) = 0.99474 and mAP50-95(M) = 0.94689 which I find pretty good and don't suggest issues with training.
![image](https://github.com/user-attachments/assets/de3d3e9d-de8a-470b-93a1-5295de5568b9)
![image](https://github.com/user-attachments/assets/6a514dbb-118d-470d-802c-0eb4a10f28d7)

**Comment by Y-T-G** [comment]
> bag is only partially detected and I find it very suspicious that the detection ends in sharp corners when there is nothing to suggest objects should stop here.

The masks cropped using the bounding boxes. They both are predicted separately. So that's normal behavior to have masks cropped at the edges when the mask is exceeding the box boundary.

**Comment by rob-safi** [comment]
The Bounding Box MAP is also very high (better then the mask MAP).
There are a couple examples in the validation set that have this issue (see image below). I guess from what you are saying the mask is not the problem but the bbox is. Does that mean I should try and increase the bbox loss weight during training?


![image](https://github.com/user-attachments/assets/21d4efd6-f4d2-4819-b76e-35c2a12a7de8)

**Comment by Y-T-G** [comment]
> The Bounding Box MAP is also very high (better then the mask MAP).
> There are a couple examples in the validation set that have this issue (see image below). I guess from what you are saying the mask is not the problem but the bbox is. Does that mean I should try and increase the bbox loss weight during training?
> 
> 
> ![image](https://github.com/user-attachments/assets/21d4efd6-f4d2-4819-b76e-35c2a12a7de8)
> 

In this image, it seems like receptive field issue. You can try P6 model.

**Comment by rob-safi** [comment]
Interesting, do you know if there is a pretrained yolo v11 with P6 model I could use?

**Comment by Y-T-G** [comment]
No

**Comment by JXNICHOLAS** [comment]
Hi,

Have you found a solution? I am experiencing the same issue here. This issue started to occur with the newer version of Ultarluytics after I upgraded my computer system.

**Comment by glenn-jocher** [comment]
Hi, the issue might be related to changes in the newer version or your updated system environment. Please ensure you're using the latest Ultralytics version and confirm your dependencies (e.g., Python, PyTorch, CUDA) are correctly installed and compatible. If the issue persists, providing more details or examples can help us assist further.

**Comment by JXNICHOLAS** [comment]
I found increasing the detection image size to around 2500 would alleviate the issue in my case.

**Comment by glenn-jocher** [comment]
Thank you for sharing your findings. Increasing the detection image size can indeed improve object detection performance for large-scale images, as it allows the model to better capture details. It's great to hear this resolved the issue in your case!

**Comment by Edwin-Koh1** [comment]
bump also having this issue as others, and the effect is more prominent on small rounded/curved objects than larger ones
